hphp/util/asm-x64-intelxed.h

```diff
@@ -0,0 +1,1066 @@
+    xedInstrRR(XED_ICLASS_ADDSD, srcdest, src);
+  }
+  void mulsd(RegXMM src, RegXMM srcdest) {
+    xedInstrRR(XED_ICLASS_MULSD, srcdest, src);
+  }
+  void subsd(RegXMM src, RegXMM srcdest) {
+    xedInstrRR(XED_ICLASS_SUBSD, srcdest, src);
+  }
+  void pxor(RegXMM src, RegXMM srcdest) {
+    xedInstrRR(XED_ICLASS_PXOR, srcdest, src);
+  }
+  void cvtsi2sd(Reg64 src, RegXMM dest) {
+    xedInstrRR(XED_ICLASS_CVTSI2SD, src, dest);
+  }
+  void cvtsi2sd(MemoryRef m, RegXMM dest) {
+    xedInstrMR(XED_ICLASS_CVTSI2SD, m, dest);
+  }
+  void ucomisd(RegXMM l, RegXMM r) {
+    xedInstrRR(XED_ICLASS_UCOMISD, l, r);
+  }
+  void sqrtsd(RegXMM src, RegXMM dest) {
+    xedInstrRR(XED_ICLASS_SQRTSD, dest, src);
+  }
+
+  void divsd(RegXMM src, RegXMM srcdest) {
+    xedInstrRR(XED_ICLASS_DIVSD, srcdest, src);
+  }
+  void cvttsd2siq(RegXMM src, Reg64 dest) {
+    xedInstrRR(XED_ICLASS_CVTTSD2SI, src, dest);
+  }
+
+private:
+  // XED conditional jump conversion functions
+#define CC_TO_XED_ARRAY(xed_instr) {                            \
+    XED_ICLASS_##xed_instr##O,    /*CC_O                  */    \
+    XED_ICLASS_##xed_instr##NO,   /*CC_NO                 */    \
+    XED_ICLASS_##xed_instr##B,    /*CC_B, CC_NAE          */    \
+    XED_ICLASS_##xed_instr##NB,   /*CC_AE, CC_NB, CC_NC   */    \
+    XED_ICLASS_##xed_instr##Z,    /*CC_E, CC_Z            */    \
+    XED_ICLASS_##xed_instr##NZ,   /*CC_NE, CC_NZ          */    \
+    XED_ICLASS_##xed_instr##BE,   /*CC_BE, CC_NA          */    \
+    XED_ICLASS_##xed_instr##NBE,  /*CC_A, CC_NBE          */    \
+    XED_ICLASS_##xed_instr##S,    /*CC_S                  */    \
+    XED_ICLASS_##xed_instr##NS,   /*CC_NS                 */    \
+    XED_ICLASS_##xed_instr##P,    /*CC_P                  */    \
+    XED_ICLASS_##xed_instr##NP,   /*CC_NP                 */    \
+    XED_ICLASS_##xed_instr##L,    /*CC_L, CC_NGE          */    \
+    XED_ICLASS_##xed_instr##NL,   /*CC_GE, CC_NL          */    \
+    XED_ICLASS_##xed_instr##LE,   /*CC_LE, CC_NG          */    \
+    XED_ICLASS_##xed_instr##NLE   /*CC_G, CC_NLE          */    \
+  }
+
+  ALWAYS_INLINE
+  xed_iclass_enum_t ccToXedJump(ConditionCode c) {
+    assertx(c != CC_None);
+    static const xed_iclass_enum_t jumps[] = CC_TO_XED_ARRAY(J);
+    return jumps[(int)c];
+  }
+
+  ALWAYS_INLINE
+  xed_iclass_enum_t ccToXedSetCC(int c) {
+    assertx(c != -1);
+    static const xed_iclass_enum_t setccs[] = CC_TO_XED_ARRAY(SET);
+    return setccs[c];
+  }
+
+  ALWAYS_INLINE
+  xed_iclass_enum_t ccToXedCMov(ConditionCode c) {
+    assertx(c != CC_None);
+    static const xed_iclass_enum_t cmovs[] = CC_TO_XED_ARRAY(CMOV);
+    return cmovs[(int)c];
+  }
+
+  // XED emit functions
+
+#define DECLARE_UNUSED(type, name)  type name; static_cast<void>(name)
+
+  ALWAYS_INLINE
+  uint32_t xedEmitImpl(xed_iclass_enum_t instr, CodeAddress destination,
+                       std::function<void(xed_encoder_instruction_t*)>
```

Comment:
Unless there's a very strong reason to use `std::function<>`, we use templates to take lambdas:
```c++
template<typename L>
uint32_t xedEmitImpl(xed_iclass_enum_t instr, CodeAddress destination,
                     L xedFunc) {
```
It's much more concise, and we don't have to worry about the compiler not being able to properly eliminate the `std::function` overhead.

Reply:
Addressed in https://github.com/facebook/hhvm/pull/8164/commits/6923adab7c8900c24fbe55f12632909549f13627